subsys/bluetooth/services/ots_oacp_internal.h

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef BT_GATT_OTS_OACP_H_
+#define BT_GATT_OTS_OACP_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <sys/types.h>
+#include <zephyr/types.h>
+#include <bluetooth/gatt.h>
+#include <bluetooth/services/ots.h>
+
+/**@brief Types of Object Action Control Point Procedures. */
+enum bt_gatt_ots_oacp_proc_type {
+	/** Create object.*/
+	BT_GATT_OTS_OACP_PROC_CREATE        = 0x01,
+	/** Delete object.*/
+	BT_GATT_OTS_OACP_PROC_DELETE        = 0x02,
+	/** Calculate Checksum.*/
+	BT_GATT_OTS_OACP_PROC_CHECKSUM_CALC = 0x03,
+	/** Execute Object.*/
+	BT_GATT_OTS_OACP_PROC_EXECUTE       = 0x04,
+	/** Read object.*/
+	BT_GATT_OTS_OACP_PROC_READ          = 0x05,
+	/** Write object.*/
+	BT_GATT_OTS_OACP_PROC_WRITE         = 0x06,
+	/** Abort object.*/
+	BT_GATT_OTS_OACP_PROC_ABORT         = 0x07,
+	/** Procedure response.*/
+	BT_GATT_OTS_OACP_PROC_RESP          = 0x60
+};
+
+/**@brief Object Action Control Point return codes. */
+enum bt_gatt_ots_oacp_res_code {
+	/** Success.*/
+	BT_GATT_OTS_OACP_RES_SUCCESS        = 0x01,
+	/** Not supported*/
+	BT_GATT_OTS_OACP_RES_OPCODE_NOT_SUP = 0x02,
+	/** Invalid parameter*/
+	BT_GATT_OTS_OACP_RES_INV_PARAM      = 0x03,
+	/** Insufficient resources.*/
+	BT_GATT_OTS_OACP_RES_INSUFF_RES     = 0x04,
+	/** Invalid object.*/
+	BT_GATT_OTS_OACP_RES_INV_OBJ        = 0x05,
+	/** Channel unavailable.*/
+	BT_GATT_OTS_OACP_RES_CHAN_UNAVAIL   = 0x06,
+	/** Unsupported procedure.*/
+	BT_GATT_OTS_OACP_RES_UNSUP_TYPE     = 0x07,
+	/** Procedure not permitted.*/
+	BT_GATT_OTS_OACP_RES_NOT_PERMITTED  = 0x08,
+	/** Object locked.*/
+	BT_GATT_OTS_OACP_RES_OBJ_LOCKED     = 0x09,
+	/** Operation Failed.*/
+	BT_GATT_OTS_OACP_RES_OPER_FAILED    = 0x0A
+};
+
+/* Object Action Control Point procedure definition. */
+struct bt_gatt_ots_oacp_proc {
```

Comment:
When using net_buf we usually don't need to have a this kind of struct with type and union of all possible packet types struct.
Instead we have just a simple struct, and pull from the buffer. You can have a look at hci command and events handling to what I mean (hci_core.c).

Reply:
I see, but it helps me to pass the data to dedicated functions inside OACP write callback and divide OACP processing into separate stages:
- ``oacp_command_decode`` (packet decoding)
- ``oacp_command_len_verify`` (length verification - ATT length error takes precedence before OTS errors)
- ``oacp_proc_validate`` (validation - checks if command is accepted by OTS)

In case of HCI Core, the HCI event type is identified, passed to the appropriate handler and processed there in a type-wise manner. Here, I use a stage-wise approach for all types:
- Initial validation
- Command decoding
- Command length verification
- OTS command validation (can it be executed?)
- Sending indication